### PR TITLE
[bazel] enable sccache+nvcc in CI

### LINKED ISSUE #79348: [bazel] build doesn't use sccache
### 🐛 Describe the bug

A build with sccache used to take 15 minutes according to @malfet .
Right now it take 65-70 minutes. This probably indicates that sccache is not used.
However this also could be because previously GPU build was not enabled and no it is.

This needs a quick investigation.

### Versions

master

### PR DESCRIPTION
**Pull request by vors**
Fixes #79348

This change is mostly focused on enabling nvcc+sccache in the PyTorch CI.

Along the way we had to do couple tweaks:
1.  Split the rules_cc from the rules_cuda that embeeded them before. This is needed in order to apply a different patch to the rules_cc compare to the one that rules_cuda does by default. This is in turn needed because we need to workaround an nvcc behavior where it doesn't send `-iquote xxx` to the host compiler, but it does send `-isystem xxx`. So we workaround this problem with (ab)using `-isystem` instead. Without it we are getting errors like `xxx` is not found.

2. Workaround bug in bazel https://github.com/bazelbuild/bazel/issues/10167 that prevents us from using a straightforward and honest `nvcc` sccache wrapper. Instead we generate ad-hock bazel specific nvcc wrapper that has internal knowledge of the relative bazel paths to local_cuda. This allows us to workaround the issue with CUDA symlinks. Without it we are getting `undeclared inclusion(s) in rule` all over the place for CUDA headers.

## Test plan

Green CI build https://github.com/pytorch/pytorch/actions/runs/4267147180/jobs/7428431740

Note that now it says "CUDA" in the sccache output

```
+ sccache --show-stats
Compile requests                    9784
Compile requests executed           6726
Cache hits                          6200
Cache hits (C/C++)                  6131
Cache hits (CUDA)                     69
Cache misses                         519
Cache misses (C/C++)                 201
Cache misses (CUDA)                  318
Cache timeouts                         0
Cache read errors                      0
Forced recaches                        0
Cache write errors                     0
Compilation failures                   0
Cache errors                           7
Cache errors (C/C++)                   7
Non-cacheable compilations             0
Non-cacheable calls                 2893
Non-compilation calls                165
Unsupported compiler calls             0
Average cache write                0.116 s
Average cache read miss           23.722 s
Average cache read hit             0.057 s
Failed distributed compilations        0
```

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -3,6 +3,19 @@ workspace(name = "pytorch")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 load("//tools/rules:workspace.bzl", "new_patched_local_repository")
 
+http_archive(
+    name = "rules_cc",
+    sha256 = "cb8ce8a25464b2a8536450971ad1b45ee309491c1f5e052a611b9e249cfdd35d",
```
Comment (kit1980) [inline WORKSPACE:8]: I think we don't want to use checksums now, see https://github.com/pytorch/pytorch/pull/95039

Comment (huydhn) [reply WORKSPACE:8]: Now the new linter becomes useful, being able to flag the line :)

File: .ci/pytorch/common_utils.sh [hunk 2]
```diff
@@ -95,6 +95,23 @@ function get_bazel() {
...
 }
 
+# This function is bazel specific because of the bug
+# in the bazel that requires some special paths massaging
+# as a workaround. See
+# https://github.com/bazelbuild/bazel/issues/10167
+function install_sccache_nvcc_for_bazel() {
+  sudo mv /usr/local/cuda/bin/nvcc /usr/local/cuda/bin/nvcc-real
+
+  echo '#!/bin/sh' | sudo tee /usr/local/cuda/bin/nvcc
```
Comment (huydhn) [inline .ci/pytorch/common_utils.sh:105]: May be cat the script here to a file, then sudo mv to `/usr/local/cuda/bin/nvcc` in one go to make this a bit easier to read?

**Review by vors — COMMENTED** [review]
Addressed the comments, thank you for the quick turn-around!

**Review by huydhn — APPROVED** [review]
LGTM!

**Comment by huydhn** [comment]
All linter failures come from SPACES linter.  Could you try to exclude `tools/rules_cc/cuda_support.patch` from that linter in https://github.com/pytorch/pytorch/blob/master/.lintrunner.toml#L367?  This should help pass the check.

**Pull request closed without merging**